File: tests/qt_types_standalone/rust/src/qpoint.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Gerhard de Clercq <gerhard.declercq@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QPoint;

#[cxx::bridge]
mod qpoint_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/include/qpoint.h");

        type QPoint = cxx_qt_lib::QPoint;
    }

    extern "Rust" {
        fn construct_qpoint() -> QPoint;
        fn read_qpoint(p: &QPoint) -> bool;
        fn clone_qpoint(p: &QPoint) -> QPoint;
    }
}

fn construct_qpoint() -> QPoint {
    QPoint::new(2, 4)
}

fn read_qpoint(p: &QPoint) -> bool {
    p.x() == 2 && p.y() == 4
}

fn clone_qpoint(p: &QPoint) -> QPoint {
    p.clone()
}

File: crates/cxx-qt-lib/src/core/qpoint.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qpoint.h"

#include <cxx-qt-lib/assertion_utils.h>

// QPoint has "int" members - xp and yp
// Rust represents these as 32-bit integer types.
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qpoint.h?h=v5.15.6-lts-lgpl#n271
//
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qpoint.h?h=v6.2.4#n313
assert_alignment_and_size(QPoint, {
  ::std::int32_t a0;
  ::std::int32_t a1;
});

static_assert(::std::is_trivially_copyable<QPoint>::value);
